controlplane/kubeadm/internal/failure_domain.go

```diff
@@ -50,15 +50,32 @@ func (f failureDomainAggregations) Swap(i, j int) {
 	f[i], f[j] = f[j], f[i]
 }
 
-// PickMost returns the failure domain with the most number of machines.
-func PickMost(failureDomains clusterv1.FailureDomains, machines FilterableMachineCollection) *string {
+// PickMost picks a machine among machines that has the failure domain with most control plane machines on.
+// After getting an ordered failure domain list based on number of control plane machines, return the machine that matches the first failure domain in the list.
+func PickMost(c *ControlPlane, machines FilterableMachineCollection) *string {
+	// orderDescending sorts failure domains according to all  control plane machines
+	fds := orderDescending(c.Cluster.Status.FailureDomains.FilterControlPlane(), c.Machines)
+	for _, fd := range fds {
+		for _, m := range machines {
```

Comment:
Goin back to this point
> Can we look at fd.count>0 instead
> > Yes, trying to find the first matching machine.

As far as I understand, what we are doing here is to check if a machine exists in the failure domain, because we are always returning a failure domain.
In that case, it is not required to loop through the machines here because this already happened in orderDescending and fd.Counts reports which fd has machines.
Is this right or I'm missing something?

Reply:
Reading this code a little bit more, it seems to say: `always return the first failure domain that has at least one machine in it`

Reply:
Yes, so there's one edge case where it will return nil, if all the failure domains are empty.

```go
if len(fds) > 0 {
  if fds[0].count > 0 {
    return &fds[0].id
  }
}
```

would do the same thing except doesn't handle the odd case of wanting to return a FD if the only machine we have is not in a FD

Reply:
> As far as I understand, what we are doing here is to check if a machine exists in the failure domain, because we are always returning a failure domain.
We get failure domain ordering looking at all control-plane machines. Whereas here, machines can be subset of control-plane machines.

So to better explain, fds are calculated from ownedMachines and a machine is selected out of selectedMachines:
```
func (r *KubeadmControlPlaneReconciler) scaleDownControlPlane(
	ctx context.Context,
	cluster *clusterv1.Cluster,
	kcp *controlplanev1.KubeadmControlPlane,
	ownedMachines internal.FilterableMachineCollection,
	selectedMachines internal.FilterableMachineCollection,
	controlPlane *internal.ControlPlane,
) 
```

We need to do the ordering according to owned machines, because we want to pick a fd with most ownedMachines, which also exist as a fd in selectedMachines.

Reply:
We are planning to remove `selectedMachines` though, so this is just going to need to change again

Reply:
Moved around things a bit. scaleDownControlPlane is now using selectMachine (renamed to selectandMarkMachine) function used in upgrade. They were doing same things.